### PR TITLE
faster indices for FlatData with lazy operations

### PR DESCRIPTION
**Pull request by AntonioCarta**
Lazy indices for FlatData to support faster operations.

### CODE REVIEW

**Comment by AntonioCarta** [comment]
@lrzpellegrini do you have any suggestion on how to check for performance (runtime) regression with automated tests?

**Comment by lrzpellegrini** [comment]
Everything seems ok!

As for the performance, it's hard to check them in an uncontrolled environment (such as GH Actions), where the processor may change. Our best shot is to run performance checks on ContinualAI servers. There are ways to tell GitHub to run actions on external runners, but I don't know if it is hard to set up.

As for the PR, I wonder if we can further improve performance by storing the eager indices in PyTorch or NumPy tensors.

File: avalanche/benchmarks/utils/flat_data.py [hunk 1]
```diff
@@ -44,6 +44,66 @@
...
+            whenever we generate them from the lazy sequence.
+            """
+            self._offset = offset
+
+        self._known_length = known_length
+
+    def _to_eager(self):
+        if self._eager_list is not None:
+            return
+        self._eager_list = [el + self._offset for el in self._lazy_sequence]
```
Comment (lrzpellegrini) [inline avalanche/benchmarks/utils/flat_data.py:84]: Maybe we can further optimize using NumPy both as a storage for _eager_list and both to optimize summing the _offset to each index.

**Comment by AntonioCarta** [comment]
> As for the PR, I wonder if we can further improve performance by storing the eager indices in PyTorch or NumPy tensors.

Maybe. Most of the cost was the concatenation of large lists for each concat operation. It should be more efficient in numpy.
For example, I added a `LazyRange` which should help avoid the "eagerification" step in some common cases.

I think we can speed this up further in the future, either pushing more on the lazyness or using numpy tensors.

**Comment by AntonioCarta** [comment]
I will leave further improvements as future work.

File: avalanche/benchmarks/utils/flat_data.py (hunk 1 above)
Comment (AntonioCarta) [reply avalanche/benchmarks/utils/flat_data.py:84]: This is useful if we do it every time we offset (there are many places). Here most cost is due to the full iteration of the iteration. Not sure numpy would help a lot. But in general, converting everything to tensors may speed up general subsets and concat ops.

**Pull request merged**